router/throttler/adaptiveAlgorithmCounter/algorithm.go

```diff
@@ -0,0 +1,102 @@
+package adaptivethrottlercounter
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+)
+
+type timer struct {
+	frequency    time.Duration
+	limitReached bool
+	mu           sync.Mutex
+	limitSet     bool
+	cancel       context.CancelFunc
+}
+
+type Adaptive struct {
+	shortTimer               *timer
+	longTimer                *timer
+	decreaseLimitPercentage  *config.Reloadable[int64]
+	increaseChangePercentage *config.Reloadable[int64]
+}
+
+func New(config *config.Config) *Adaptive {
+	shortTimeFrequency := config.GetDuration("Router.throttler.adaptive.shortTimeFrequency", 5, time.Second)
+	longTimeFrequency := config.GetDuration("Router.throttler.adaptive.longTimeFrequency", 15, time.Second)
```

Comment:
These should be proportional to the configured window of the destination. 
If the destination has a window of `60s` and the algorithm has a window of `1s`, the result would be to limit the rate too aggressively causing large starvation periods.